tests/test_tensornet.py

```diff
@@ -700,3 +700,187 @@ def test_hermitian(self, theta, phi, varphi, tol):
 
         assert np.allclose(res, expected, atol=tol, rtol=0)
         
+
+class TestSample:
+    """Tests that samples are properly calculated."""
+
+    def test_sample_dimensions(self, tensornet_device_2_wires):
+        """Tests if the samples returned by the sample function have
+        the correct dimensions
+        """
+
+        # Explicitly resetting is necessary as the internal
+        # state is set to None in __init__ and only properly
+        # initialized during reset
+        tensornet_device_2_wires.reset()
+
+        tensornet_device_2_wires.apply('RX', wires=[0], par=[1.5708])
+        tensornet_device_2_wires.apply('RX', wires=[1], par=[1.5708])
+
+        tensornet_device_2_wires.shots = 10
+        s1 = tensornet_device_2_wires.sample('PauliZ', [0], [])
+        assert np.array_equal(s1.shape, (10,))
+
+        tensornet_device_2_wires.shots = 12
+        s2 = tensornet_device_2_wires.sample('PauliZ', [1], [])
+        assert np.array_equal(s2.shape, (12,))
+
+        tensornet_device_2_wires.shots = 17
+        s3 = tensornet_device_2_wires.sample('CZ', [0, 1], [])
+        assert np.array_equal(s3.shape, (17,))
+
+    def test_sample_values(self, tensornet_device_2_wires, tol):
+        """Tests if the samples returned by sample have
+        the correct values
+        """
+
+        # Explicitly resetting is necessary as the internal
+        # state is set to None in __init__ and only properly
+        # initialized during reset
+        tensornet_device_2_wires.reset()
+
+        tensornet_device_2_wires.apply('RX', wires=[0], par=[1.5708])
+
+        s1 = tensornet_device_2_wires.sample('PauliZ', [0], [])
+
+        # s1 should only contain 1 and -1, which is guaranteed if
+        # they square to 1
+        assert np.allclose(s1**2, 1, atol=tol, rtol=0)
+
+
+@pytest.mark.parametrize("theta, phi, varphi", list(zip(THETA, PHI, VARPHI)))
+class TestTensorSample:
+    """Test tensor expectation values"""
+
+    def test_paulix_pauliy(self, theta, phi, varphi, monkeypatch, tol):
+        """Test that a tensor product involving PauliX and PauliY works correctly"""
+        dev = qml.device("expt.tensornet", wires=3, shots=10000)
+        dev.reset()
+        dev.apply("RX", wires=[0], par=[theta])
+        dev.apply("RX", wires=[1], par=[phi])
+        dev.apply("RX", wires=[2], par=[varphi])
+        dev.apply("CNOT", wires=[0, 1], par=[])
+        dev.apply("CNOT", wires=[1, 2], par=[])
+
+        with monkeypatch.context() as m:
```

Comment:
Maybe this was copied from elsewhere, but what is the `monkeypatch` doing here?

Reply:
I believe that it mocks out the ``numpy.random.choice`` function, such that there are no random values returned, but instead the list of values to sample from and the probabilities are returned (first and third arguments to ``numpy.random.choice``)

This can be done this way, as we assume that the external library (``numpy``) takes care of testing the functionality of this function